packages/modular-scripts/src/esbuild-scripts/plugins/workerFactoryPlugin.ts

```diff
@@ -0,0 +1,141 @@
+          const relativePath =
+            './' + path.relative(process.cwd(), workerAbsolutePath);
+
+          return {
+            path: relativePath,
+            namespace: 'web-worker',
+          };
+        },
+      );
+
+      build.onLoad({ filter: /.*/, namespace: 'web-worker' }, async (args) => {
+        const relativePath = path.relative(getModularRoot(), args.path);
+
+        // Build the worker file with the same format, target and definitions of the bundle
+        try {
+          const result = await esbuild.build({
+            format: build.initialOptions.format,
+            target: build.initialOptions.target,
+            define: build.initialOptions.define,
+            minify: build.initialOptions.minify,
+            entryPoints: [args.path],
+            plugins: [createExtensionAllowlistPlugin()],
+            bundle: true,
+            write: false,
+          });
+
+          // Store the file in the build cache for later use, since we need to emit a file and trampoline it transparently to the user
+          workerBuildCache.set(relativePath, result);
+
+          // Trampoline the worker within the bundle, to avoid CORS errors
+          return {
+            contents: `
+                  // Web worker bundled by worker-factory-plugin, mimicking the Worker constructor
+                  import workerUrl from 'worker-url:${relativePath}';
+                  
+                  const workerPath = new URL(workerUrl, import.meta.url);
+                  const importSrc = 'import "' + workerPath + '";';
+
+                  const blob = new Blob([importSrc], {
+                    type: "text/javascript",
+                  });
+
+                  export default class {
+                    constructor() {
+                      return new Worker(URL.createObjectURL(blob), { type: "module" });
+                    }
+                  }
+                `,
+          };
+        } catch (e) {
+          console.error('Error building worker script:', e);
+        }
+      });
+
+      build.onResolve({ filter: /worker-url:.*/ }, (args) => {
+        return {
+          path: args.path.slice('worker-url:'.length),
+          namespace: 'worker-url',
+        };
+      });
+
+      build.onLoad({ filter: /.*/, namespace: 'worker-url' }, (args) => {
+        const result = workerBuildCache.get(args.path);
+        if (result) {
+          const { outputFiles } = result;
+          if (outputFiles?.length === 1) {
+            const outputFile = outputFiles[0];
+            return {
+              contents: outputFile.contents,
+              loader: 'file',
+            };
+          } else {
+            throw new Error(`Could not read output files`);
+          }
+        } else {
+          throw new Error(`Could not find result for ${args.path}`);
+        }
+      });
+    },
+  };
+
+  return plugin;
+}
+
+function createExtensionAllowlistPlugin(
```

Comment:
Can we split this into a seperate plugin file for consistency with the others?

Reply:
👍🏻 done